File: app/src/main/java/com/lenskart/trendinggits/data/RepoAdapter.java
package com.lenskart.trendinggits.data;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;
import com.lenskart.trendinggits.R;

import java.util.ArrayList;

public class RepoAdapter extends RecyclerView.Adapter<RepoAdapter.ViewHolder> implements AdapterView.OnItemClickListener, Filterable {
    private final ArrayList<Repo> dataSet = new ArrayList<>();
    private ArrayList<Repo> repoOriginalFullList;
    private OnUserItemClickedListener mOnUserItemClickedListener;

    public RepoAdapter(ArrayList<Repo> mUserList, OnUserItemClickedListener onUserItemClickedListener) {
        this.dataSet.addAll(mUserList);
        repoOriginalFullList = mUserList;
        this.mOnUserItemClickedListener = onUserItemClickedListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.repo_list, parent, false);
        return new ViewHolder(view, mOnUserItemClickedListener);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        RequestOptions requestOptions = new RequestOptions();
        requestOptions = requestOptions.transform(new CenterCrop(), new RoundedCorners(16));

        Glide.with(holder.view.getContext())
                .load("https://github.com/" + dataSet.get(position).getOwnerAvatar() + ".png")
                .centerCrop()
                .apply(requestOptions)
                .placeholder(R.drawable.ic_launcher_foreground)
                .into(holder.repoAvatarImg);
        holder.repoNameTxt.setText(dataSet.get(position).getRepoName());
        holder.repoLinkTxt.setText(dataSet.get(position).getRepoLink());
        holder.repoLangTxt.setText(dataSet.get(position).getLanguage());
        holder.repoForkCountTxt.setText(dataSet.get(position).getTotalForks());
        holder.repoStarTxt.setText(dataSet.get(position).getTotalStars());
        holder.todayRepoStarTxt.setText(dataSet.get(position).getTodayStars());
        if (dataSet.get(position).isSelected()) {
            holder.cardViewItem.setCardBackgroundColor(Color.GRAY);
        } else {
            holder.cardViewItem.setCardBackgroundColor(Color.WHITE);
        }
    }

    @Override
    public int getItemCount() {
        return dataSet.size();
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {

    }

    @Override
    public Filter getFilter() {
        return searchedFilter;
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        View view;
        TextView repoNameTxt;
        CardView cardViewItem;
        ImageView repoAvatarImg;
        TextView repoLinkTxt;
        TextView repoLangTxt;
        TextView repoForkCountTxt;
        TextView repoStarTxt;
        TextView todayRepoStarTxt;
        OnUserItemClickedListener onUserItemClickedListener;

        public ViewHolder(@NonNull View itemView, OnUserItemClickedListener onUserItemClickedListener) {
            super(itemView);
            view = itemView;
            cardViewItem = view.findViewById(R.id.cardViewItem);
            repoNameTxt = view.findViewById(R.id.repoNameTxt);
            repoAvatarImg = view.findViewById(R.id.repoAvatar);
            repoLinkTxt = view.findViewById(R.id.repoLinkTxt);
            repoLangTxt = view.findViewById(R.id.repoLangTxt);
            repoForkCountTxt = view.findViewById(R.id.repoForkCountTxt);
            repoStarTxt = view.findViewById(R.id.repoStarTxt);
            todayRepoStarTxt = view.findViewById(R.id.todayRepoStarTxt);
            this.onUserItemClickedListener = onUserItemClickedListener;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {

            if(dataSet.get(getAbsoluteAdapterPosition()).isSelected()) {
                cardViewItem.setCardBackgroundColor(Color.WHITE);
                dataSet.get(getAbsoluteAdapterPosition()).setSelected(false);
            }else {
                cardViewItem.setCardBackgroundColor(Color.GRAY);
                dataSet.get(getAbsoluteAdapterPosition()).setSelected(true);
            }

        }
    }

    public interface OnUserItemClickedListener {
        void onUserItemClicked(int itemId);
    }


    private final Filter searchedFilter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            ArrayList<Repo> filteredList = new ArrayList<>();
            if (constraint == null || constraint.length() == 0) {
                filteredList.addAll(repoOriginalFullList);
            } else {
                String filterPattern = constraint.toString().toLowerCase().trim();
                for (Repo item : repoOriginalFullList) {
                    if (item.getRepoName().toLowerCase().contains(filterPattern)) {
                        filteredList.add(item);
                    }
                }
            }
            FilterResults results = new FilterResults();
            results.values = filteredList;
            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            dataSet.clear();
            dataSet.addAll((ArrayList) results.values);
            notifyDataSetChanged();
        }
    };
}

File: app/src/main/java/com/lenskart/trendinggits/data/RepoAdapter.kt
package com.lenskart.trendinggits.data

import android.graphics.Color
import androidx.recyclerview.widget.RecyclerView
import android.widget.AdapterView.OnItemClickListener
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import com.lenskart.trendinggits.R
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.load.resource.bitmap.CenterCrop
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.bumptech.glide.Glide
import androidx.cardview.widget.CardView
import java.util.*
import kotlin.collections.ArrayList

class RepoAdapter(
    mUserList: ArrayList<Repo>,
    onUserItemClickedListener: OnUserItemClickedListener
) : RecyclerView.Adapter<RepoAdapter.ViewHolder>(), OnItemClickListener, Filterable {
    private val dataSet = ArrayList<Repo>()
    private var repoOriginalFullList: ArrayList<Repo> = ArrayList()
    private val mOnUserItemClickedListener: OnUserItemClickedListener
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.repo_list, parent, false)
        return ViewHolder(view, mOnUserItemClickedListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        var requestOptions = RequestOptions()
        requestOptions = requestOptions.transform(CenterCrop(), RoundedCorners(16))
        Glide.with(holder.view.context)
            .load("https://github.com/" + dataSet[position].ownerAvatar + ".png")
            .centerCrop()
            .apply(requestOptions)
            .placeholder(R.drawable.ic_launcher_foreground)
            .into(holder.repoAvatarImg)
        holder.repoNameTxt.text = dataSet[position].repoName
        holder.repoLinkTxt.text = dataSet[position].repoLink
        holder.repoLangTxt.text = dataSet[position].language
        holder.repoForkCountTxt.text = dataSet[position].totalForks
        holder.repoStarTxt.text = dataSet[position].totalStars
        holder.todayRepoStarTxt.text = dataSet[position].todayStars
        if (dataSet[position].isSelected) {
            holder.cardViewItem.setCardBackgroundColor(Color.GRAY)
        } else {
            holder.cardViewItem.setCardBackgroundColor(Color.WHITE)
        }
    }

    override fun getItemCount(): Int {
        return dataSet.size
    }

    override fun onItemClick(adapterView: AdapterView<*>?, view: View, i: Int, l: Long) {}
    override fun getFilter(): Filter {
        return searchedFilter
    }

    inner class ViewHolder(var view: View, onUserItemClickedListener: OnUserItemClickedListener) :
        RecyclerView.ViewHolder(
            view
        ), View.OnClickListener {
        var repoNameTxt: TextView
        var cardViewItem: CardView
        var repoAvatarImg: ImageView
        var repoLinkTxt: TextView
        var repoLangTxt: TextView
        var repoForkCountTxt: TextView
        var repoStarTxt: TextView
        var todayRepoStarTxt: TextView
        var onUserItemClickedListener: OnUserItemClickedListener
        override fun onClick(view: View) {
            if (dataSet[absoluteAdapterPosition].isSelected) {
                cardViewItem.setCardBackgroundColor(Color.WHITE)
                dataSet[absoluteAdapterPosition].isSelected = false
            } else {
                cardViewItem.setCardBackgroundColor(Color.GRAY)
                dataSet[absoluteAdapterPosition].isSelected = true
            }
            onUserItemClickedListener.onUserItemClicked(repoOriginalFullList.indexOf(dataSet[absoluteAdapterPosition]))
        }

        init {
            cardViewItem = view.findViewById(R.id.cardViewItem)
            repoNameTxt = view.findViewById(R.id.repoNameTxt)
            repoAvatarImg = view.findViewById(R.id.repoAvatar)
            repoLinkTxt = view.findViewById(R.id.repoLinkTxt)
            repoLangTxt = view.findViewById(R.id.repoLangTxt)
            repoForkCountTxt = view.findViewById(R.id.repoForkCountTxt)
            repoStarTxt = view.findViewById(R.id.repoStarTxt)
            todayRepoStarTxt = view.findViewById(R.id.todayRepoStarTxt)
            this.onUserItemClickedListener = onUserItemClickedListener
            itemView.setOnClickListener(this)
        }
    }

    interface OnUserItemClickedListener {
        fun onUserItemClicked(itemId: Int)
    }

    private val searchedFilter: Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val filteredList = ArrayList<Repo>()
            if (constraint == null || constraint.isEmpty()) {
                filteredList.addAll(repoOriginalFullList)
            } else {
                val filterPattern =
                    constraint.toString().lowercase(Locale.getDefault()).trim { it <= ' ' }
                for (item in repoOriginalFullList) {
                    if (item.repoName.lowercase(Locale.getDefault()).contains(filterPattern)) {
                        filteredList.add(item)
                    }
                }
            }
            val results = FilterResults()
            results.values = filteredList
            return results
        }

        override fun publishResults(constraint: CharSequence?, results: FilterResults) {
            dataSet.clear()
            dataSet.addAll(results.values as ArrayList<Repo>)
            notifyDataSetChanged()
        }
    }

    init {
        dataSet.addAll(mUserList)
        repoOriginalFullList = mUserList
        mOnUserItemClickedListener = onUserItemClickedListener
    }
}